Compiler crashes exposing internal errors if a comment with no code following it is encountered

See the following program:
```
foo =
    # TODO

main = 42
```

### Expected behaviour

Of course we are missing any statement in the `foo` method. But it should just print some syntax error, or something.

### Actual behaviour

Right now it crashes with:
```
java.lang.NullPointerException: Cannot invoke "org.enso.compiler.core.ir.Expression.transformExpressions(scala.PartialFunction)" because "ir" is null
        at org.enso.runtime/org.enso.compiler.pass.resolve.DocumentationComments$.org$enso$compiler$pass$resolve$DocumentationComments$$resolveExpression(DocumentationComments.scala:83)
        at org.enso.runtime/org.enso.compiler.pass.resolve.DocumentationComments$$anonfun$org$enso$compiler$pass$resolve$DocumentationComments$$resolveExpression$1.applyOrElse(DocumentationComments.scala:87)
        at org.enso.runtime/org.enso.compiler.pass.resolve.DocumentationComments$$anonfun$org$enso$compiler$pass$resolve$DocumentationComments$$resolveExpression$1.applyOrElse(DocumentationComments.scala:83)
        at org.enso.runtime/scala.runtime.AbstractPartialFunction.apply(AbstractPartialFunction.scala:35)
        at org.enso.runtime/org.enso.compiler.core.ir.Expression.transformExpressions(Expression.scala:22)
        at org.enso.runtime/org.enso.compiler.core.ir.Expression.transformExpressions$(Expression.scala:18)
        at org.enso.runtime/org.enso.compiler.core.ir.Expression$Block.transformExpressions(Expression.scala:58)
        at org.enso.runtime/org.enso.compiler.pass.resolve.DocumentationComments$.org$enso$compiler$pass$resolve$DocumentationComments$$resolveExpression(DocumentationComments.scala:83)
        at org.enso.runtime/org.enso.compiler.pass.resolve.DocumentationComments$.resolveDefinition(DocumentationComments.scala:167)
        at org.enso.runtime/org.enso.compiler.pass.resolve.DocumentationComments$.$anonfun$resolveModule$5(DocumentationComments.scala:211)
        at org.enso.runtime/scala.collection.immutable.List.map(List.scala:246)
        at org.enso.runtime/org.enso.compiler.pass.resolve.DocumentationComments$.resolveModule(DocumentationComments.scala:211)
        at org.enso.runtime/org.enso.compiler.pass.resolve.DocumentationComments$.runModule(DocumentationComments.scala:58)
        at org.enso.runtime/org.enso.compiler.pass.PassManager.$anonfun$runPassesOnModule$2(PassManager.scala:101)
        at org.enso.runtime/scala.collection.LinearSeqOps.foldLeft(LinearSeq.scala:183)
        at org.enso.runtime/scala.collection.LinearSeqOps.foldLeft$(LinearSeq.scala:179)
        at org.enso.runtime/scala.collection.immutable.List.foldLeft(List.scala:79)
        at org.enso.runtime/org.enso.compiler.pass.PassManager.runPassesOnModule(PassManager.scala:92)
        at org.enso.runtime/org.enso.compiler.Compiler.recognizeBindings(Compiler.scala:799)
        at org.enso.runtime/org.enso.compiler.Compiler.uncachedParseModule(Compiler.scala:631)
        at org.enso.runtime/org.enso.compiler.Compiler.parseModule(Compiler.scala:586)
        at org.enso.runtime/org.enso.compiler.Compiler.$anonfun$runCompilerPipeline$1(Compiler.scala:251)
        at org.enso.runtime/org.enso.compiler.Compiler.$anonfun$runCompilerPipeline$1$adapted(Compiler.scala:249)
        at org.enso.runtime/scala.collection.immutable.List.foreach(List.scala:333)
        at org.enso.runtime/org.enso.compiler.Compiler.runCompilerPipeline(Compiler.scala:249)
        at org.enso.runtime/org.enso.compiler.Compiler.go$1(Compiler.scala:233)
        at org.enso.runtime/org.enso.compiler.Compiler.runInternal(Compiler.scala:240)
        at org.enso.runtime/org.enso.compiler.Compiler.run(Compiler.scala:131)
        at org.enso.runtime/org.enso.interpreter.runtime.Module.compile(Module.java:393)
        at org.enso.runtime/org.enso.interpreter.runtime.Module.compileScope(Module.java:329)
        at org.enso.runtime/org.enso.interpreter.node.ProgramRootNode.execute(ProgramRootNode.java:66)
        at org.graalvm.truffle.runtime/com.oracle.truffle.runtime.OptimizedCallTarget.executeRootNode(OptimizedCallTarget.java:745)
        at org.graalvm.truffle.runtime/com.oracle.truffle.runtime.OptimizedCallTarget.profiledPERoot(OptimizedCallTarget.java:669)
        at org.graalvm.truffle.runtime/com.oracle.truffle.runtime.OptimizedCallTarget.callBoundary(OptimizedCallTarget.java:602)
        at org.graalvm.truffle.runtime/com.oracle.truffle.runtime.OptimizedCallTarget.doInvoke(OptimizedCallTarget.java:586)
        at org.graalvm.truffle.runtime/com.oracle.truffle.runtime.OptimizedCallTarget.callIndirect(OptimizedCallTarget.java:519)
        at org.graalvm.truffle.runtime/com.oracle.truffle.runtime.OptimizedCallTarget.call(OptimizedCallTarget.java:500)
        at org.graalvm.truffle/com.oracle.truffle.polyglot.PolyglotContextImpl.eval(PolyglotContextImpl.java:1683)
        at org.graalvm.truffle/com.oracle.truffle.polyglot.PolyglotContextDispatch.eval(PolyglotContextDispatch.java:60)
        at org.graalvm.polyglot/org.graalvm.polyglot.Context.eval(Context.java:402)
        at org.enso.IsolatedClassLoader//org.enso.polyglot.PolyglotContext.evalModule(PolyglotContext.scala:32)
        at org.enso.IsolatedClassLoader//org.enso.runner.Main.runSingleFile(Main.java:836)
        at org.enso.IsolatedClassLoader//org.enso.runner.Main.handleRun(Main.java:740)
        at org.enso.IsolatedClassLoader//org.enso.runner.Main.mainEntry(Main.java:1093)
        at org.enso.IsolatedClassLoader//org.enso.runner.Main.lambda$launch$20(Main.java:1431)
        at org.enso.IsolatedClassLoader//org.enso.runner.Main.withProfiling(Main.java:1183)
        at org.enso.IsolatedClassLoader//org.enso.runner.Main.launch(Main.java:1427)
        at org.enso.IsolatedClassLoader//org.enso.runner.Main.launch(Main.java:1384)
        at org.enso.IsolatedClassLoader//org.enso.runner.Main.main(Main.java:1001)
        at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
        at java.base/java.lang.reflect.Method.invoke(Method.java:580)
        at org.enso.runtime/org.enso.EngineRunnerBootLoader.main(EngineRunnerBootLoader.java:46)
[ERROR] [2024-08-20T14:33:30.643] [enso.org.enso.compiler.Compiler] Contents of module X:\NBO\enso\example.enso: foo =
    # TODO

main = 42


[ERROR] [2024-08-20T14:33:30.651] [org.enso.runner.Main] Error during execution
java.io.IOException: org.graalvm.polyglot.PolyglotException: java.lang.NullPointerException: Cannot invoke "org.enso.compiler.core.ir.Expression.transformExpressions(scala.PartialFunction)" because "ir" is null
        at org.enso.IsolatedClassLoader//org.enso.runner.Main.withProfiling(Main.java:1187)
        at org.enso.IsolatedClassLoader//org.enso.runner.Main.launch(Main.java:1427)
        at org.enso.IsolatedClassLoader//org.enso.runner.Main.launch(Main.java:1384)
        at org.enso.IsolatedClassLoader//org.enso.runner.Main.main(Main.java:1001)
        at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
        at java.base/java.lang.reflect.Method.invoke(Method.java:580)
        at org.enso.runtime/org.enso.EngineRunnerBootLoader.main(EngineRunnerBootLoader.java:46)
Caused by: org.graalvm.polyglot.PolyglotException: java.lang.NullPointerException: Cannot invoke "org.enso.compiler.core.ir.Expression.transformExpressions(scala.PartialFunction)" because "ir" is null
        at org.enso.runtime/org.enso.compiler.pass.resolve.DocumentationComments$.org$enso$compiler$pass$resolve$DocumentationComments$$resolveExpression(DocumentationComments.scala:83)
        at org.enso.runtime/org.enso.compiler.pass.resolve.DocumentationComments$$anonfun$org$enso$compiler$pass$resolve$DocumentationComments$$resolveExpression$1.applyOrElse(DocumentationComments.scala:87)
        at org.enso.runtime/org.enso.compiler.pass.resolve.DocumentationComments$$anonfun$org$enso$compiler$pass$resolve$DocumentationComments$$resolveExpression$1.applyOrElse(DocumentationComments.scala:83)
        at org.enso.runtime/scala.runtime.AbstractPartialFunction.apply(AbstractPartialFunction.scala:35)
        at org.enso.runtime/org.enso.compiler.core.ir.Expression.transformExpressions(Expression.scala:22)
        at org.enso.runtime/org.enso.compiler.core.ir.Expression.transformExpressions$(Expression.scala:18)
        at org.enso.runtime/org.enso.compiler.core.ir.Expression$Block.transformExpressions(Expression.scala:58)
        at org.enso.runtime/org.enso.compiler.pass.resolve.DocumentationComments$.org$enso$compiler$pass$resolve$DocumentationComments$$resolveExpression(DocumentationComments.scala:83)
        at org.enso.runtime/org.enso.compiler.pass.resolve.DocumentationComments$.resolveDefinition(DocumentationComments.scala:167)
        at org.enso.runtime/org.enso.compiler.pass.resolve.DocumentationComments$.$anonfun$resolveModule$5(DocumentationComments.scala:211)
        at org.enso.runtime/scala.collection.immutable.List.map(List.scala:246)
        at org.enso.runtime/org.enso.compiler.pass.resolve.DocumentationComments$.resolveModule(DocumentationComments.scala:211)
        at org.enso.runtime/org.enso.compiler.pass.resolve.DocumentationComments$.runModule(DocumentationComments.scala:58)
        at org.enso.runtime/org.enso.compiler.pass.PassManager.$anonfun$runPassesOnModule$2(PassManager.scala:101)
        at org.enso.runtime/scala.collection.LinearSeqOps.foldLeft(LinearSeq.scala:183)
        at org.enso.runtime/scala.collection.LinearSeqOps.foldLeft$(LinearSeq.scala:179)
        at org.enso.runtime/scala.collection.immutable.List.foldLeft(List.scala:79)
        at org.enso.runtime/org.enso.compiler.pass.PassManager.runPassesOnModule(PassManager.scala:92)
        at org.enso.runtime/org.enso.compiler.Compiler.recognizeBindings(Compiler.scala:799)
        at org.enso.runtime/org.enso.compiler.Compiler.uncachedParseModule(Compiler.scala:631)
        at org.enso.runtime/org.enso.compiler.Compiler.parseModule(Compiler.scala:586)
        at org.enso.runtime/org.enso.compiler.Compiler.ensureParsed(Compiler.scala:689)
        at org.enso.runtime/org.enso.compiler.Compiler.ensureParsed(Compiler.scala:678)
        at org.enso.runtime/org.enso.compiler.phase.ImportResolver.go$1(ImportResolver.scala:152)
        at org.enso.runtime/org.enso.compiler.phase.ImportResolver.mapImports(ImportResolver.scala:165)
        at org.enso.runtime/org.enso.compiler.Compiler.liftedTree1$1(Compiler.scala:456)
        at org.enso.runtime/org.enso.compiler.Compiler.runImportsAndExportsResolution(Compiler.scala:455)
        at org.enso.runtime/org.enso.compiler.Compiler.$anonfun$runCompilerPipeline$2(Compiler.scala:273)
        at org.enso.runtime/scala.collection.immutable.List.flatMap(List.scala:293)
        at org.enso.runtime/org.enso.compiler.Compiler.runCompilerPipeline(Compiler.scala:268)
        at org.enso.runtime/org.enso.compiler.Compiler.go$1(Compiler.scala:233)
        at org.enso.runtime/org.enso.compiler.Compiler.runInternal(Compiler.scala:240)
        at org.enso.runtime/org.enso.compiler.Compiler.run(Compiler.scala:131)
        at org.enso.runtime/org.enso.interpreter.runtime.Module.compile(Module.java:393)
        at org.enso.runtime/org.enso.interpreter.runtime.Module.compileScope(Module.java:329)
        at org.enso.runtime/org.enso.interpreter.node.ProgramRootNode.execute(ProgramRootNode.java:66)
        at <enso>.(Unknown)
        at org.graalvm.polyglot/org.graalvm.polyglot.Context.eval(Context.java:402)
        at org.enso.IsolatedClassLoader//org.enso.polyglot.PolyglotContext.evalModule(PolyglotContext.scala:32)
        at org.enso.IsolatedClassLoader//org.enso.runner.Main.runSingleFile(Main.java:836)
        at org.enso.IsolatedClassLoader//org.enso.runner.Main.handleRun(Main.java:740)
        at org.enso.IsolatedClassLoader//org.enso.runner.Main.mainEntry(Main.java:1093)
        at org.enso.IsolatedClassLoader//org.enso.runner.Main.lambda$launch$20(Main.java:1431)
        at org.enso.IsolatedClassLoader//org.enso.runner.Main.withProfiling(Main.java:1183)
        ... 6 common frames omitted
Command failed with an error: java.io.IOException: org.graalvm.polyglot.PolyglotException: java.lang.NullPointerException: Cannot invoke "org.enso.compiler.core.ir.Expression.transformExpressions(scala.PartialFunction)" because "ir" is null
```

### Comment

For comparison the following program:
```
foo =

main = 42
```
actually executes without any errors or warnings and returns `42`. So it seems to be the comment that causes the issue.

--- a/engine/runtime-integration-tests/src/test/java/org/enso/compiler/ErrorCompilerTest.java
+++ b/engine/runtime-integration-tests/src/test/java/org/enso/compiler/ErrorCompilerTest.java
@@ -4,6 +4,7 @@ import static org.junit.Assert.assertEquals;
 import static org.junit.Assert.assertTrue;
 
 import org.enso.compiler.core.ir.Empty;
+import org.enso.compiler.core.ir.Expression;
 import org.enso.compiler.core.ir.Location;
 import org.enso.compiler.core.ir.Module;
 import org.enso.compiler.core.ir.expression.errors.Syntax;
@@ -578,6 +579,19 @@ public class ErrorCompilerTest extends CompilerTest {
 
     var method = (Method) ir.bindings().apply(0);
     assertTrue(method.body() instanceof Empty);
+  }
+
+  @Test
+  public void testBodyWithComment() throws Exception {
+    var ir = parse("""
+    main =
+        # comment
+    """);
+
+    var method = (Method) ir.bindings().apply(0);
+    var body = (Expression.Block) method.body();
+    assertTrue(body.expressions().isEmpty());
+    assertTrue(body.returnValue() instanceof Empty);
   }
 
   @Test

--- a/engine/runtime-integration-tests/src/test/scala/org/enso/interpreter/test/instrument/RuntimeServerTest.scala
+++ b/engine/runtime-integration-tests/src/test/scala/org/enso/interpreter/test/instrument/RuntimeServerTest.scala
@@ -7463,6 +7463,54 @@ class RuntimeServerTest
     )
   }
 
+  it should "run methods with empty body ending with comment" in {
+    val contextId  = UUID.randomUUID()
+    val requestId  = UUID.randomUUID()
+    val moduleName = "Enso_Test.Test.Main"
+    val code =
+      """foo =
+        |    # TODO
+        |
+        |main =
+        |    foo
+        |""".stripMargin.linesIterator.mkString("\n")
+
+    val mainFile = context.writeMain(code)
+
+    // create context
+    context.send(Api.Request(requestId, Api.CreateContextRequest(contextId)))
+    context.receive shouldEqual Some(
+      Api.Response(requestId, Api.CreateContextResponse(contextId))
+    )
+
+    // Set sources for the module
+    context.send(
+      Api.Request(requestId, Api.OpenFileRequest(mainFile, code))
+    )
+    context.receive shouldEqual Some(
+      Api.Response(Some(requestId), Api.OpenFileResponse)
+    )
+
+    // push main
+    context.send(
+      Api.Request(
+        requestId,
+        Api.PushContextRequest(
+          contextId,
+          Api.StackItem.ExplicitCall(
+            Api.MethodPointer(moduleName, moduleName, "main"),
+            None,
+            Vector()
+          )
+        )
+      )
+    )
+    context.receiveNIgnoreStdLib(2) should contain theSameElementsAs Seq(
+      Api.Response(requestId, Api.PushContextResponse(contextId)),
+      context.executionComplete(contextId)
+    )
+  }
+
   it should "support file edit notification with IdMap" in {
     val contextId  = UUID.randomUUID()
     val requestId  = UUID.randomUUID()

--- a/engine/runtime-parser/src/main/java/org/enso/compiler/core/TreeToIr.java
+++ b/engine/runtime-parser/src/main/java/org/enso/compiler/core/TreeToIr.java
@@ -1019,20 +1019,21 @@ final class TreeToIr {
           }
           last = translateExpression(expr, false);
         }
-        // If the block ended in a documentation node without an expression, last may be null;
-        // the return value of the block is a doc comment.
-        // (This is to match the behavior of AstToIr; after the parser transition, we should probably
-        // ignore the orphaned documentation and return the last actual expression in the block.)
-        if (last == null && expressions.size() > 0) {
-          last = expressions.get(expressions.size() - 1);
-          expressions.remove(expressions.size() - 1);
+        var locationWithANewLine = getIdentifiedLocation(body, 0, 0, null);
+        if (last == null) {
+          if (expressions.isEmpty()) {
+            last = new Empty(locationWithANewLine, meta());
+          } else {
+            last = expressions.get(expressions.size() - 1);
+            expressions.remove(expressions.size() - 1);
+          }
         }
         var list = CollectionConverters.asScala(expressions.iterator()).toList();
-        var locationWithANewLine = getIdentifiedLocation(body, 0, 0, null);
-        if (last != null && last.location().isDefined()
+        if (last != null
+            && last.location().isDefined()
             && last.location().get().end() != locationWithANewLine.get().end()) {
-          var patched = new Location(last.location().get().start(),
-              locationWithANewLine.get().end() - 1);
+          var patched =
+              new Location(last.location().get().start(), locationWithANewLine.get().end() - 1);
           var id = IdentifiedLocation.create(patched, last.location().get().id());
           last = last.setLocation(Option.apply(id));
         }

